File: frontend/src/features/analytics/legacyLineBarCodes.ts
/** Mirrors LegacyLineBarCodes on the backend: maps a pre-grouping fused Line/Bar code from an old bookmarked URL to (grouping, code). */
export const LEGACY_LINE_BAR_CODES: Record<
    string,
    { grouping: string; code: string }
> = {
    "Line Chart": { grouping: "None", code: "Raw Values" },
    "Aggregated Sum": { grouping: "Exact", code: "Sum" },
    "Cumulative Sum": { grouping: "Exact", code: "Cumulative Sum" },
    Daily: { grouping: "Daily", code: "Sum" },
    Weekly: { grouping: "Weekly", code: "Sum" },
    Monthly: { grouping: "Monthly", code: "Sum" },
    Yearly: { grouping: "Yearly", code: "Sum" },
    "Count Bar Chart": { grouping: "Exact", code: "Count" },
    "Sum Bar Chart": { grouping: "Exact", code: "Sum" },
    "Average Bar Chart": { grouping: "Exact", code: "Average" },
    "Daily Bar Chart": { grouping: "Daily", code: "Sum" },
    "Weekly Bar Chart": { grouping: "Weekly", code: "Sum" },
    "Monthly Bar Chart": { grouping: "Monthly", code: "Sum" },
    "Yearly Bar Chart": { grouping: "Yearly", code: "Sum" },
};

export function resolveLegacyCalculation(
    grouping: string | null | undefined,
    code: string | null | undefined,
): { grouping: string | null; code: string | null } {
    if (grouping || !code || !(code in LEGACY_LINE_BAR_CODES))
        return { grouping: grouping ?? null, code: code ?? null };
    const mapped = LEGACY_LINE_BAR_CODES[code];
    return { grouping: mapped.grouping, code: mapped.code };
}

File: backend/src/Operum.Model/Constants/Analytics/LegacyLineBarCodes.cs
namespace Operum.Model.Constants.Analytics
{
    // Maps retired fused Line/Bar codes (e.g. "Daily") to their (Grouping, Code) replacement,
    // for the AddWidgetGrouping migration and old bookmarked Explore URLs.
    public static class LegacyLineBarCodes
    {
        public static readonly Dictionary<string, (string Grouping, string Code)> Map = new()
        {
            // Line
            ["Line Chart"] = (AnalyticGroupings.None, AnalyticCodes.RawValues),
            ["Aggregated Sum"] = (AnalyticGroupings.Exact, AnalyticCodes.Sum),
            ["Cumulative Sum"] = (AnalyticGroupings.Exact, AnalyticCodes.CumulativeSum),
            ["Daily"] = (AnalyticGroupings.Daily, AnalyticCodes.Sum),
            ["Weekly"] = (AnalyticGroupings.Weekly, AnalyticCodes.Sum),
            ["Monthly"] = (AnalyticGroupings.Monthly, AnalyticCodes.Sum),
            ["Yearly"] = (AnalyticGroupings.Yearly, AnalyticCodes.Sum),

            // Bar
            ["Count Bar Chart"] = (AnalyticGroupings.Exact, AnalyticCodes.Count),
            ["Sum Bar Chart"] = (AnalyticGroupings.Exact, AnalyticCodes.Sum),
            ["Average Bar Chart"] = (AnalyticGroupings.Exact, AnalyticCodes.Average),
            ["Daily Bar Chart"] = (AnalyticGroupings.Daily, AnalyticCodes.Sum),
            ["Weekly Bar Chart"] = (AnalyticGroupings.Weekly, AnalyticCodes.Sum),
            ["Monthly Bar Chart"] = (AnalyticGroupings.Monthly, AnalyticCodes.Sum),
            ["Yearly Bar Chart"] = (AnalyticGroupings.Yearly, AnalyticCodes.Sum),
        };

        public static (string? Grouping, string Code) Resolve(string? grouping, string code)
        {
            if (!string.IsNullOrEmpty(grouping))
                return (grouping, code);

            return Map.TryGetValue(code, out var pair) ? (pair.Grouping, pair.Code) : (grouping, code);
        }
    }
}
